jmix-pivottable/pivottable-flowui-kit/src/main/java/io/jmix/pivottableflowui/kit/component/JmixPivotTable.java

```diff
@@ -0,0 +1,1033 @@
+
+    protected static final String DATA_ITEM_ID_PROPERTY_NAME = "$k";
+
+    protected DataProvider<DataItem, ?> dataProvider;
+    protected Registration dataProviderItemSetChangeRegistration;
+    protected Registration cellClickJsRegistration;
+
+    protected PivotTableOptions options;
+    protected JmixPivotTableSerializer serializer;
+
+    protected volatile boolean clientReady;
+    protected final List<PendingJsFunction> functions = new ArrayList<>();
+
+    protected StateTree.ExecutionRegistration synchronizeOptionsExecution;
+    protected StateTree.ExecutionRegistration synchronizeItemsExecution;
+
+    protected Map<String, Registration> eventRegistrations = new HashMap<>();
+
+    public JmixPivotTable() {
+        initComponent();
+    }
+
+    public DataProvider<DataItem, ?> getDataProvider() {
+        return dataProvider;
+    }
+
+    /**
+     * Sets a data provider to the pivot table or replaces existing one.
+     * Using the {@link DataProvider}, the component can easily retrieve data from any data source.
+     *
+     * @param dataProvider data source to set
+     */
+    public void setDataProvider(DataProvider<DataItem, ?> dataProvider) {
+        if (Objects.equals(dataProvider, this.dataProvider)) {
+            return;
+        }
+
+        this.dataProvider = dataProvider;
+        onDataProviderChange();
+    }
+
+    /**
+     * Init component data with varargs of {@link DataItem}
+     * @param dataItems
+     */
+    public void setData(DataItem... dataItems) {
+        if (dataItems != null) {
+            setDataProvider(new PivotTableListDataSet<>(Arrays.asList(dataItems)));
+        }
+    }
+
+    /**
+     * Init component data with the list of {@link DataItem}
+     * @param dataItems
+     */
+    public void setItems(List<DataItem> dataItems) {
+        setDataProvider(new PivotTableListDataSet<>(dataItems));
+    }
+
+    /**
+     * Adds a listener to the pivot table refresh events. Fired only for editable PivotTable.
+     *
+     * @param listener a listener to add
+     * @return subscription
+     */
+    public Registration addRefreshEventListener(ComponentEventListener<PivotTableRefreshEvent> listener) {
+        return getEventBus().addListener(PivotTableRefreshEvent.class, listener);
+    }
+
+    /**
+     * Adds a listener to the pivot table cell click events. Fired only for table
+     * renderers (table, heatmap, table barchart, col heatmap, row heatmap).
+     *
+     * @param listener a listener to add
+     * @return subscription
+     */
+    public Registration addCellClickListener(ComponentEventListener<PivotTableCellClickEvent> listener) {
+        if (cellClickJsRegistration == null) {
+            cellClickJsRegistration = getEventBus().addListener(PivotTableJsCellClickEvent.class, this::onCellClick);
+        }
+
+        Registration registration = getEventBus().addListener(PivotTableCellClickEvent.class, listener);
+        return () -> {
```

Comment:
We used such contractions because we wrap Vaadin components in Classic UI. Here you can directrly subsribe on dom events. Also, take a look at Kanban as an inspiration.